File: src/api/scheduler.ts
import { tauriInvoke } from "@api/tauri";

export type SchedulerTaskType = "Restart" | "Backup" | "Command";

export interface ScheduledTask {
  id: string;
  name: string;
  task_type: SchedulerTaskType;
  cron_expression: string;
  command?: string | null;
  enabled: boolean;
  last_run?: string | null;
  next_run?: string | null;
}

export const schedulerApi = {
  async getAllTasks(): Promise<ScheduledTask[]> {
    return tauriInvoke<ScheduledTask[]>("get_all_tasks");
  },

  async createTask(params: {
    name: string;
    taskType: SchedulerTaskType;
    cronExpression: string;
    command?: string | null;
  }): Promise<ScheduledTask> {
    return tauriInvoke<ScheduledTask>("create_task", {
      name: params.name,
      task_type: params.taskType,
      cron_expression: params.cronExpression,
      command: params.command ?? null,
    });
  },

  async updateTask(params: {
    id: string;
    name: string;
    taskType: SchedulerTaskType;
    cronExpression: string;
    command?: string | null;
    enabled: boolean;
  }): Promise<ScheduledTask> {
    return tauriInvoke<ScheduledTask>("update_task", {
      id: params.id,
      name: params.name,
      task_type: params.taskType,
      cron_expression: params.cronExpression,
      command: params.command ?? null,
      enabled: params.enabled,
    });
  },

  async deleteTask(id: string): Promise<void> {
    return tauriInvoke<void>("delete_task", { id });
  },

  async toggleTask(id: string): Promise<ScheduledTask> {
    return tauriInvoke<ScheduledTask>("toggle_task", { id });
  },

  async runTaskNow(id: string): Promise<void> {
    return tauriInvoke<void>("run_task_now", { id });
  },
};

File: src-tauri/src/commands/scheduler.rs
use crate::models::scheduler::{ScheduledTask, TaskType};
use crate::services::global;

fn service() -> &'static crate::services::scheduler_service::SchedulerService {
    global::scheduler_service()
}

#[tauri::command]
pub fn get_all_tasks() -> Vec<ScheduledTask> {
    service().get_all_tasks()
}

#[tauri::command]
pub fn create_task(
    name: String,
    task_type: TaskType,
    cron_expression: String,
    command: Option<String>,
) -> Result<ScheduledTask, String> {
    let task = ScheduledTask {
        id: String::new(),
        name,
        task_type,
        cron_expression,
        command,
        enabled: true,
        last_run: None,
        next_run: None,
    };
    service().add_task(task)
}

#[tauri::command]
pub fn update_task(
    id: String,
    name: String,
    task_type: TaskType,
    cron_expression: String,
    command: Option<String>,
    enabled: bool,
) -> Result<ScheduledTask, String> {
    let task = ScheduledTask {
        id,
        name,
        task_type,
        cron_expression,
        command,
        enabled,
        last_run: None,
        next_run: None,
    };
    service().update_task(task)
}

#[tauri::command]
pub fn delete_task(id: String) -> Result<(), String> {
    service().remove_task(&id)
}

#[tauri::command]
pub fn toggle_task(id: String) -> Result<ScheduledTask, String> {
    service().toggle_task(&id)
}

#[tauri::command]
pub fn run_task_now(id: String) -> Result<(), String> {
    service().run_task_now(&id)
}
